### ISSUE TITLE
Allow explicit listing of sources and deactivation of glob

### ISSUE THREAD

**Issue by GPMueller** [Trick-17/clang-build#8]
Specifying folders instead of files is more sensible, but sometimes a complete glob is too much. Therefore it would be good to have the options to
- manually specify additional source files outside of the given folders
- deactivate globbing of folders (per target)

**Comment by NOhs** [comment]
I disagree. This sounds like trying to adapt the build system to some specific existing project layouts. Do you have a good argument, why this would help a lot of people? Sticking to the folder structure with all projects would certainly help finding your way around in other projects.

**Comment by GPMueller** [comment]
I agree it's good practice to separate sources of different targets but I don't want to assume that it's always the right choice.
Also, what about unit tests? I would place them as single cpps into a single 'tests' folder and define one target per test (though we may think about making this easier somehow).

Lastly, if other people use bad project structures it will be very hard to integrate them as dependencies.
It does not need to be easy but I belive to some extent it should be possible.

**Comment by NOhs** [comment]
If you think about how `catch` works, you can separate tests into as many cpps as you like. Dependencies can either be libraries against which you link or header only libraries, which also works. I think it is too early to abandon structure ;).

**Comment by GPMueller** [comment]
I would agree that a single test executable is usually the right choice, but for targets which do not use `catch` (big mistake!), individual test executables may be the preferrable way (or what did you mean?).

Another question is: which of these is better
1. force people to only use third-party projects which have this quite specific folder structure
2. allow for most third-party projects to be used, but also allow "bad" project structures

I believe people would always try to do it the easy way (i.e. the "good" project structure) and only use the non-glob target setup if necessary (e.g. projects they have little control over).

**Comment by GPMueller** [comment]
While  explicit vs. implicit mostly depends on taste, there is another reason to allow explicit listing of sources, namely that scaling builds to large and complex codebases seems to work more reliably with explicit listing. Though I have not experienced this, a lot of people seem to share this opinion, see for example  [this reddit comment](https://www.reddit.com/r/cpp/comments/524844/recommend_a_build_system/d7itluo/) or the discussions under [this reddit thread](https://www.reddit.com/r/cpp/comments/881vt4/cmake_311_released/)

**Comment by GPMueller** [comment]
Actually, since the code uses globbing on specified directories anyways, specifying glob patterns instead of source directories would make sense and we could replace
```
[targetname.sources]
include_directories = ["something"]
source_directories = ["src/a", "src/b"]
```
(*note: we should probably drop the `[targetname.sources]`*) with
```
include_directories = ["something"]
sources = ["src/*"]
```
where `sources = ["src/a/*", "src/b/*"]` would still work as well. This would give people the option to be explicit, if they desire, by writing
```
include_directories = ["something"]
sources = ["src/a/a.cpp", "src/b/b.cpp"]
```

**Comment by GPMueller** [comment]
We should take a closer look at how glob patterns work in Python, as it seems that "src/*" will not yield the desired result, especially if it should be recursive.

Excluded paths could be implemented as it's currently done on `feature-sources-patterns` with cf6685dcc45484c8a9b88235ec99232a7c66ef5a with `target.sources.include_exclude` and `target.sources.source_exclude`. This might be more convenient than placing exclude patterns into the `target.sources.include_directories` and `target.sources.source_directories` glob patterns.

We could also distinguish between folders and patterns by testing if pathlib can resolve the corresponding string.

**Comment by GPMueller** [comment]
One more thing to consider: if `include_directories` are not used in their current form, they cannot be used for the include flags.

**Comment by GPMueller** [comment]
Implemented with 35e8cfc7d953fcea71f7f9667354b4bf09194c2d on `feature-sources-patterns` (see PR #63).
